cli/azd/internal/telemetry/storage_test.go

```diff
@@ -0,0 +1,243 @@
+package telemetry
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+	"github.com/benbjohnson/clock"
+	"github.com/stretchr/testify/assert"
+)
+
+// The tests in this file intentionally interacts with the filesystem (important implementation detail).
+// As such, it might be susceptible to filesystem related failures and also general slowness.
+
+const fileExtension = ".itm"
+
+type tempFolder struct {
+	name string
+}
+
+func newTempDir(t *testing.T, name string) tempFolder {
+	dirname, err := os.MkdirTemp("", name)
+	assert.NoError(t, err)
+	return tempFolder{dirname}
+}
+
+func (t *tempFolder) close() {
+	_ = os.RemoveAll(t.name)
+}
+
+func TestNewStorageQueue(t *testing.T) {
+	folder := filepath.Join(os.TempDir(), "azdnewstg")
+	defer os.RemoveAll(folder)
+
+	t.Run("CreatesFolder", func(t *testing.T) {
+		err := os.RemoveAll(folder)
+		assert.NoError(t, err)
+
+		storage, err := NewStorageQueue(folder, fileExtension)
+		assert.NoError(t, err)
+		assert.DirExists(t, storage.folder)
+	})
+
+	t.Run("HandlesExistingFolder", func(t *testing.T) {
+		err := os.MkdirAll(folder, osutil.PermissionDirectory)
+		assert.NoError(t, err)
+
+		storage, err := NewStorageQueue(folder, fileExtension)
+		assert.NoError(t, err)
+		assert.DirExists(t, storage.folder)
+	})
+}
+
+func TestFifoQueue(t *testing.T) {
+	dir := newTempDir(t, "azdq")
+	defer dir.close()
+
+	messages := []string{
+		"Message1",
+		"Message2",
+		"Message3",
+	}
+
+	storage := setupStorageQueue(t, dir)
+	// Queue 3 items, asserting each one is stored
+	item1 := enqueueAndAssert(storage, messages[0], t)
+	item2 := enqueueAndAssert(storage, messages[1], t)
+	item3 := enqueueAndAssert(storage, messages[2], t)
+
+	// Remove a non-head entry. This is only possible
+	// since we Peek the head as we queue each item.
+	err := storage.Remove(item1)
+	assert.NoError(t, err)
+
+	// Assert head unchanged
+	item, err := storage.Peek()
+	assert.NoError(t, err)
+	assert.Equal(t, messages[2], string(item.Message()))
```

Comment:
OK so this is indeed a LIFO queue, which I am not sure is what we really want for telemetry...

Reply:
Pretty embarrassing to get the test assertion wrong too, but it's fixed.